### PR TITLE
Validator caches index

### PR DESCRIPTION
**Pull request by terencechain**
In worst case, a validator requests its index 4 times on per slot basis. From getting assignment to proposing block, to submit unaggregation attestation and finally submit aggregation request.

This PR improves upon that by saving validator index and conveniently updates them during `UpdateAssignments`. This cuts the worst case to 1 time.

Tested run time 👍

### CODE REVIEW

File: validator/client/validator.go [hunk 1]
```diff
@@ -251,6 +253,9 @@ func (v *validator) UpdateAssignments(ctx context.Context, slot uint64) error {
 	v.assignments = resp
 	// Only log the full assignments output on epoch start to be less verbose.
 	if slot%params.BeaconConfig().SlotsPerEpoch == 0 {
+		v.pubKeyToIDLock.Lock()
+		defer v.pubKeyToIDLock.Unlock()
+
 		for _, assignment := range v.assignments.ValidatorAssignment {
 			// TODO(4379): Make validator index part of the assignment respond.
 			res, err := v.validatorClient.ValidatorIndex(ctx, &pb.ValidatorIndexRequest{PublicKey: assignment.PublicKey})
```
Comment (prestonvanloon) [inline validator/client/validator.go]: Reuse cache here?

File: validator/client/validator_aggregate.go [hunk 2]
```diff
@@ -94,17 +94,15 @@ func (v *validator) waitToSlotTwoThirds(ctx context.Context, slot uint64) {
...
 
-func (v *validator) addIndicesToLog(ctx context.Context, committeeIndex uint64, pubKey []byte) error {
+func (v *validator) addIndicesToLog(ctx context.Context, committeeIndex uint64, pubKey [48]byte) error {
 	v.attLogsLock.Lock()
 	defer v.attLogsLock.Unlock()
-	res, err := v.validatorClient.ValidatorIndex(ctx, &pb.ValidatorIndexRequest{PublicKey: pubKey})
-	if err != nil {
-		return err
-	}
+	v.pubKeyToIDLock.Lock()
```
Comment (prestonvanloon) [inline validator/client/validator_aggregate.go]: RLock?

File: validator/client/validator_aggregate.go [hunk 3]
```diff
@@ -94,17 +94,15 @@ func (v *validator) waitToSlotTwoThirds(ctx context.Context, slot uint64) {
...
+	defer v.pubKeyToIDLock.Unlock()
 
 	for _, log := range v.attLogs {
 		if committeeIndex == log.data.CommitteeIndex {
-			log.aggregatorIndices = append(log.aggregatorIndices, res.Index)
+			log.aggregatorIndices = append(log.aggregatorIndices, v.pubKeyToID[pubKey])
```
Comment (prestonvanloon) [inline validator/client/validator_aggregate.go]: What if this value is missing?

File: validator/client/validator_attest.go [hunk 4]
```diff
@@ -129,22 +129,18 @@ func (v *validator) assignment(pubKey [48]byte) (*pb.AssignmentResponse_Validato
...
-func (v *validator) indexInCommittee(
-	ctx context.Context,
-	pubKey [48]byte,
-	assignment *pb.AssignmentResponse_ValidatorAssignment) (uint64, uint64, error) {
-	res, err := v.validatorClient.ValidatorIndex(ctx, &pb.ValidatorIndexRequest{PublicKey: pubKey[:]})
-	if err != nil {
-		return 0, 0, err
-	}
+func (v *validator) indexInCommittee(pubKey [48]byte, assignment *pb.AssignmentResponse_ValidatorAssignment) (uint64, uint64, error) {
+	v.pubKeyToIDLock.Lock()
```
Comment (prestonvanloon) [inline validator/client/validator_attest.go]: RLock?

File: validator/client/validator_attest_test.go [hunk 5]
```diff
@@ -130,7 +114,7 @@ func TestAttestToBlockHead_AttestsCorrectly(t *testing.T) {
 	validator.SubmitAttestation(context.Background(), 30, validatorPubKey)
 
 	aggregationBitfield := bitfield.NewBitlist(uint64(len(committee)))
-	aggregationBitfield.SetBitAt(4, true)
+	aggregationBitfield.SetBitAt(0, true)
```
Comment (prestonvanloon) [inline validator/client/validator_attest_test.go]: Why?

File: validator/client/validator_propose.go [hunk 6]
```diff
@@ -72,20 +72,16 @@ func (v *validator) ProposeBlock(ctx context.Context, slot uint64, pubKey [48]by
 		trace.Int64Attribute("numAttestations", int64(len(b.Body.Attestations))),
 	)
 
-	res, err := v.validatorClient.ValidatorIndex(ctx, &pb.ValidatorIndexRequest{PublicKey: pubKey[:]})
-	if err != nil {
-		log.WithError(err).Error("Failed to get validator index")
-		return
-	}
-
+	v.pubKeyToIDLock.Lock()
```
Comment (prestonvanloon) [inline validator/client/validator_propose.go]: RLock?

File: validator/client/validator_attest_test.go (hunk 5 above)
Comment (terencechain) [reply validator/client/validator_attest_test.go]: before we hardcoded validator index request to always return 4th validator, we no longer request validator index hence the new validator index is the first validator in the committee given the pubkey we pass in

File: validator/client/validator_aggregate.go (hunk 3 above)
Comment (terencechain) [reply validator/client/validator_aggregate.go]: Added check

**Pull request merged**